# python-tutorial
### Introduction

#### Definition
Python is a dynamic and strongly typed programming language. It employs both duck typing and gradual typing, via type hints.

Imperative, declarative (e.g., functional), and object-oriented programming styles are all supported, but internally everything in Python is an object.

Python puts a strong emphasis on code readability and (similar to Haskell) uses significant indentation to denote function, method, and class definitions.

Python was created by Guido van Rossum and first released in 1991.

#### Why Python?

1.	Readability: Python’s syntax is designed to be clear and straightforward. It reads almost like plain English, making it easy for newcomers to understand and write code.
2.	Versatility: Python is a Swiss Army knife of programming languages. You can use it for web development, data analysis, scientific computing, automation, artificial intelligence, and more.
3.	Large Community: Python boasts a vibrant and supportive community. There are countless tutorials, libraries, and forums to help you learn and solve problems.
4.	Cross-Platform: Whether you use Windows, macOS, or Linux, Python works seamlessly on all major operating systems.
5.	Rich Ecosystem: Python comes with a vast standard library that provides pre-built modules for various tasks, saving 
	you time and effort.




### Comment 
We can add comments to our Python scripts. Comments are important to make sure that you and others can understand what your code is about and do not run as Python code. They start with # tag. 







### Python as a calculator
Python is perfectly suited to do basic calculations. It can do addition, subtraction, multiplication and division.

In [5]:
# Addition
print(4+5)

# Subtraction
print(5 - 5)

# Multiplication
print(3*5)

# Division
print(10/2)


9
0
15
5.0


### Variables 

 A variable is a name that refers to a value stored in the computer's memory. It acts as a placeholder or a label, allowing programmers to access and manipulate data throughout their code. In Python, a variable allows you to refer to a value with a name. Variables are created by assigning a value to them using the equal sign (=), and their values can be changed as needed.
 

Python is dynamically typed, meaning we don't need to declare a variable's type ahead of time; the type is inferred at runtime based on the assigned value.

In [6]:
x = 10        # An integer assignment
name = "Leila" # A string assignment

# Here, x and name are variables. x holds an integer value 10, and name holds the string "Leila".

print(x)
name

10


'Leila'

In [7]:
# Create a variable savings
savings = 100

# Print out savings
print(savings)

100


#### Calculations with variables
You've now created a savings variable, so let's start saving!

Instead of calculating with the actual values, you can use variables instead. The savings variable you created in the previous exercise with a value of 100 is available to you.

How much money would you have saved four months from now, if you saved £10 each month?

In [8]:
# Create the variables monthly_savings and num_months
monthly_savings = 10
num_months = 4

# Multiply monthly_savings and num_months
new_savings = monthly_savings * num_months

# Add new_savings to your savings
total_savings = savings + new_savings

# Print total_savings
print(total_savings)


140


### Variables Type (Data Types in Python)

1. integers (int): Used for whole numbers.
2. Floating-point numbers (float): Used for numbers with decimals or in exponential form.
3. Strings (str): Used for textual data. You can use single or double quotes to build a string.
4. Booleans (bool): a type to represent logical values. It can only be True or False (the capitalization is important!).
5. Lists (list): Mutable sequences used to store collections of items (elements can be added, removed, or changed).
6. Tuples (tuple): Immutable sequences, meaning once created, their elements cannot be changed.
7. Dictionaries (dict): Mutable collections that store data as key-value pairs.
8. Sets (set): Mutable collections of unique elements used to perform common set operations.
9. Each of these types plays a crucial role in handling data in Python, enabling developers to solve a wide 

In [9]:
# Create a variable half
half = 0.5

# Create a variable intro
intro = "Hello! How are you?"

# Create a variable is_good
is_good = True 

#### Guess the type
To find out the type of a value or a variable that refers to that value, you can use the type() function. Suppose you've defined a variable a, but you forgot the type of this variable. To determine the type of a, simply execut: type(a)



In [10]:
type(half)

float

#### Operations with other types
Note: Different types behave differently in Python.

When you sum two strings, for example, you'll get different behavior than when you sum two integers or two booleans.


In [11]:
monthly_savings = 10
num_months = 12
intro = "Hello! How are you?"

# Calculate year_savings using monthly_savings and num_months
year_savings = monthly_savings * num_months

# Print the type of year_savings
print(type(year_savings))

# Assign sum of intro and intro to doubleintro
doubleintro = intro + intro

# Print out doubleintro
print(doubleintro)

<class 'int'>
Hello! How are you?Hello! How are you?


#### Type conversion
Using the + operator to paste together two strings can be very useful in building custom messages.

Suppose, for example, that you've calculated your savings want to summarize the results in a string.

To do this, you'll need to explicitly convert the types of your variables. More specifically, you'll need str(), to convert a value into a string. str(savings), for example, will convert the integer savings to a string.

Similar functions such as int(), float() and bool() will help you convert Python values into any type.

In [12]:
# Definition of savings and total_savings
savings = 100
total_savings = 150

# Fix the printout
print("I started with $" + str(savings) + " and now have $" + str(total_savings) + ". Awesome!")

# Definition of pi_string
pi_string = "3.1415926"

# Convert pi_string into float: pi_float
pi_float = float(pi_string)
pi_float

I started with $100 and now have $150. Awesome!


3.1415926

##### Arithmetic operations with different types 
Note : In Python, the boolean values True and False are treated as 1 and 0 respectively when used in arithmetic contexts. Therefore, adding True and False together results in 1 (since 1 + 0 = 1)

In [13]:
"I can add integers, like " + str(5) + " to strings."

"I said " + ("Hey " * 2) + "Hey!"

"The correct answer to this multiple choice exercise is answer number " + str(2)

True + False

1

### List
As opposed to int, bool etc., a list is a compound data type; you can group values together.

- Data Structure: A list is an ordered collection of items, which can be of different types (integers, strings, other lists, etc.).
- Mutable: Lists can be modified after their creation; items can be added, removed, or changed.
- Indexable: Each item in a list has a specific order and can be accessed using an index, starting from 0 for the first element.
- Flexible: Lists can grow or shrink dynamically as items are added or removed.
- Syntax: Defined by square brackets [], with items separated by commas, e.g., [1, 'hello', 3.14].
- Versatile Functions: Supports numerous methods like append(), remove(), sort(), and many more to manipulate the data.
- Supports Nesting: Lists can contain other lists, creating matrix-like structures or more complex data formats.

In [14]:
# area variables (in square meters)
hall = 11.25
kit = 18.0
liv = 20.0
bed = 10.75
bath = 9.50

# Create list areas
areas = [hall,kit,liv,bed,bath]
# Print areas
print(areas)

[11.25, 18.0, 20.0, 10.75, 9.5]


##### Create list with different types
A list can contain any Python type. Although it's not really common, a list can also contain a mix of Python types including strings, floats, booleans, etc.

The printout of the previous exercise wasn't really satisfying. It's just a list of numbers representing the areas, but you can't tell which area corresponds to which part of your house.

The code in the editor is the start of a solution. For some of the areas, the name of the corresponding room is already placed in front. Pay attention here! "bathroom" is a string, while bath is a variable that represents the float 9.50 you specified earlier.

In [15]:
# area variables (in square meters)
hall = 11.25
kit = 18.0
liv = 20.0
bed = 10.75
bath = 9.50

# Adapt list areas
areas = ["hallway", hall, "kitchen", kit, "living room", liv, "bedroom", bed, "bathroom", bath]

# Print areas
print(areas)


['hallway', 11.25, 'kitchen', 18.0, 'living room', 20.0, 'bedroom', 10.75, 'bathroom', 9.5]


A list can contain any Python type. But a list itself is also a Python type. That means that a list can also contain a list!

In [16]:
A= [1, 3, 4, 2] 
B=[[1, 2, 3], [4, 5, 7]] 
C= [1 + 2, "a" * 5, 3]

C

[3, 'aaaaa', 3]

#### List of lists
You'll often be dealing with a lot of data, and it will make sense to group some of this data.

Instead of creating a flat list containing strings and floats, representing the names and areas of the rooms in your house, you can create a list of lists. 

Don't get confused here: "hallway" is a string, while hall is a variable that represents the float 11.25 you specified earlier.

In [17]:
# house information as list of lists
house = [["hallway", hall],
         ["kitchen", kit],
         ["living room", liv],
         ["bedroom", bed],
         ["bathroom", bath]]

# Print out house
print(house)

# Print out the type of house
print(type(house))


[['hallway', 11.25], ['kitchen', 18.0], ['living room', 20.0], ['bedroom', 10.75], ['bathroom', 9.5]]
<class 'list'>


#### Subset and conquer
Subsetting Python lists is a piece of cake. Take the code sample below, which creates a list x and then selects "b" from it. Remember that this is the second element, so it has index 1. You can also use negative indexing.




In [18]:
x = ["a", "b", "c", "d"]
x[1]
x[-3] #same result!

'b'

So the areas list from before, containing both strings and floats. Let's add the correct code to do some Python subsetting.

In [19]:
# Print out second element from areas
print(areas[1])

# Print out last element from areas
print(areas[-1])

# Print out the area of the living room
print(areas[5])

11.25
9.5
20.0


#### Subset and calculate
After you've extracted values from a list, you can use them to perform additional calculations. Take this example, where the second and fourth element of a list x are extracted. The strings that result are pasted together using the + operator:

In [20]:

print(x[1] + x[3])

bd


In [21]:
# Sum of kitchen and bedroom area: eat_sleep_area
eat_sleep_area = areas[3] + areas[-3]

# Print the variable eat_sleep_area
print(eat_sleep_area)


28.75


#### Slicing and dicing
Selecting single values from a list is just one part of the story. It's also possible to slice your list, which means selecting multiple elements from your list. Use the following syntax:

my_list[start:end]

The start index will be included, while the end index is not.

However, it's also possible not to specify these indexes. If you don't specify the begin index, Python figures out that you want to start your slice at the beginning of your list. If you don't specify the end index, the slice will go all the way to the last element of your list. 


In [22]:
x[1:3]
x[:2]
x[2:]
x[:]

['a', 'b', 'c', 'd']

In [23]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Use slicing to create downstairs (first 6 elements)
downstairs =areas [:6]

# Use slicing to create upstairs (last 4 elements)
upstairs =areas [6:]

# Print out downstairs and upstairs
print(downstairs)
print(upstairs)

['hallway', 11.25, 'kitchen', 18.0, 'living room', 20.0]
['bedroom', 10.75, 'bathroom', 9.5]


#### Subsetting lists of lists
You saw before that a Python list can contain practically anything; even other lists! To subset lists of lists, you can use the same technique as before: square brackets.

In [24]:
x = [["a", "b", "c"],
     ["d", "e", "f"],
     ["g", "h", "i"]]
x[2][0]
x[2][:2]

['g', 'h']

x[2] results in a list, that you can subset again by adding additional square brackets.

In [25]:
house[-1][1]

9.5

#### Replace list elements
Replacing list elements is pretty easy. Simply subset the list and assign new values to the subset. You can select single elements or you can change entire list slices at once.

In [26]:
x[1] = "r"
x[2:] = ["s", "t"]
x

[['a', 'b', 'c'], 'r', 's', 't']

In [27]:
# Correct the bathroom area
areas[-1] = 10.50

# Change "living room" to "chill zone"
areas[4] = "chill zone"

#### Extend a list
If you can change elements in a list, you sure want to be able to add elements to it, right? You can use the + operator.
+ operator add all the contains of the lists into one single list.

In [28]:
x = ["a", "b", "c", "d"]
y = x + ["e", "f"]
y

['a', 'b', 'c', 'd', 'e', 'f']

In [29]:
# Create the areas list and make some changes
areas = ["hallway", 11.25, "kitchen", 18.0, "chill zone", 20.0,
         "bedroom", 10.75, "bathroom", 10.50]

# Add poolhouse data to areas, new list is areas_1

areas_1 = areas + ["poolhouse", 24.5]
# Add garage data to areas_1, new list is areas_2
areas_2 = areas_1 + ["garage", 15.45]
areas_2

['hallway',
 11.25,
 'kitchen',
 18.0,
 'chill zone',
 20.0,
 'bedroom',
 10.75,
 'bathroom',
 10.5,
 'poolhouse',
 24.5,
 'garage',
 15.45]

#### Delete list elements
Finally, you can also remove elements from your list. You can do this with the del statement:

In [30]:
del(x[1])

Pay attention here: as soon as you remove an element from a list, the indexes of the elements that come after the deleted element all cha

There was a mistake! You decide to delete 'poolhouse' and its area value from the area_2 list



In [31]:
del(areas_2[-4:-2])

In [32]:
areas_2

['hallway',
 11.25,
 'kitchen',
 18.0,
 'chill zone',
 20.0,
 'bedroom',
 10.75,
 'bathroom',
 10.5,
 'garage',
 15.45]

### Note:

The ; sign is used to place commands on the same line. The following two code chunks are equivalent:

#### Same line
command1; command2

#### Separate lines
- command1
- command2

### Inner workings of lists
How Python lists work behind the scenes:

The Python code in the script already creates a list with the name areas and a copy named areas_copy. Next, the first element in the areas_copy list is changed and the areas list is printed out. If you hit Run Code you'll see that, although you've changed areas_copy, the change also takes effect in the areas list. That's because areas and areas_copy point to the same list.

If you want to prevent changes in areas_copy from also taking effect in areas, you'll have to do a more explicit copy of the areas list. You can do this with list() or by using [:].

In [33]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Create areas_copy
areas_copy = areas

# Change areas_copy
areas_copy[0] = 5.0

# Print areas
print(areas)

[5.0, 18.0, 20.0, 10.75, 9.5]


Change the second command, that creates the variable areas_copy, such that areas_copy is an explicit copy of areas. After your edit, changes made to areas_copy shouldn't affect areas. Submit the answer to check this.

In [34]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Create areas_copy
areas_copy = list(areas)

# Change areas_copy
areas_copy[0] = 5.0

# Print areas
print(areas)

[11.25, 18.0, 20.0, 10.75, 9.5]


#### Familiar functions
Out of the box, Python offers a bunch of built-in functions to make your life easier. You already know two such functions: print() and type(). You've also used the functions str(), int(), bool() and float() to switch between data types. These are built-in functions as well.

Calling a function is easy. To get the type of 3.0 and store the output as a new variable, result, you can use the following:
result = type(3.0)

The general recipe for calling functions and saving the result to a variable is thus:
output = function_name(input)



In [35]:
# Create variables var1 and var2
var1 = [1, 2, 3, 4]
var2 = True

# Print out type of var1
print(type(var1))

# Print out length of var1
print(len(var1))

# Convert var2 to an integer: out2
out2 = int(var2)
print(out2)

<class 'list'>
4
1


Help!
Maybe you already know the name of a Python function, but you still have to figure out how to use it. Ironically, you have to ask for information about a function with another function: help(). In IPython specifically, you can also use ? before the function name.

To get help on the max() function, for example, you can use one of these calls:

help(max)
?max

In [36]:
help(pow)

Help on built-in function pow in module builtins:

pow(base, exp, mod=None)
    Equivalent to base**exp with 2 arguments or base**exp % mod with 3 arguments
    
    Some types, such as ints, are able to use a more efficient algorithm when
    invoked using the three argument form.



So, pow() takes three arguments: base, exp, and mod. base and exp are required arguments, mod is an optional argument.

#### Multiple arguments
In the previous exercise, you identified optional arguments by viewing the documentation with help(). You'll now apply this to change the behavior of the sorted() function.

Have a look at the documentation of sorted() by typing help(sorted) in the IPython Shell.

You'll see that sorted() takes three arguments: iterable, key, and reverse.

key=None means that if you don't specify the key argument, it will be None. reverse=False means that if you don't specify the reverse argument, it will be False, by default.

In the exercise below, you'll only have to specify iterable and reverse, not key. The first input you pass to sorted() will be matched to the iterable argument, but what about the second input? To tell Python you want to specify reverse without changing anything about key, you can use = to assign it a new value:

sorted(____, reverse=____)

In [37]:
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [38]:
# Create lists first and second
first = [11.25, 18.0, 20.0]
second = [10.75, 9.50]

# Paste together first and second: full
full = first + second 

# Sort full in descending order: full_sorted

full_sorted = sorted(full, reverse = True)
# Print out full_sorted

print(full_sorted)


[20.0, 18.0, 11.25, 10.75, 9.5]


#### String Methods
Strings come with a bunch of methods. Follow the instructions closely to discover some of them. If you want to discover them in more detail, you can always type help(str).

In [39]:
# string to experiment with: place
place = "poolhouse"

# Use upper() on place: place_up
place_up = place.upper()

# Print out place and place_up
print(place)
print(place_up)

# Print out the number of o's in place
print(place.count("o"))

poolhouse
POOLHOUSE
3


#### List Methods
Strings are not the only Python types that have methods associated with them. Lists, floats, integers and booleans are also types that come packaged with a bunch of useful methods. In this exercise, you'll be experimenting with:

index(), to get the index of the first element of a list that matches its input and
count(), to get the number of times an element appears in a list.
You'll be working on the list with the area of different parts of a house: areas.

In [40]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Print out the index of the element 20.0
print(areas.index(20.0))

# Print out how often 9.50 appears in areas
print(areas.count(9.50))

2
1


Most list methods will change the list they're called on. Examples are:

append(), that adds an element to the list it is called on,
remove(), that removes the first element of a list that matches the input, and
reverse(), that reverses the order of the elements in the list it is called on.
You'll be working on the list with the area of different parts of the house: areas.

In [ ]:
# Create list areas
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Use append twice to add poolhouse and garage size
areas.append(24.5)
areas.append(15.45)


# Print out areas
print(areas)

# Reverse the orders of the elements in areas
areas.reverse()

# Print out areas
print(areas)

#### Import package

For a fancy clustering algorithm, you want to find the circumference, , and area, , of a circle. When the radius of the circle is r, you can calculate  and  as:


In Python, the symbol for exponentiation is **. This operator raises the number to its left to the power of the number to its right. For example 3**4 is 3 to the power of 4 and will give 81.

To use the constant pi, you'll need the math package. A variable r is already coded in the script. Fill in the code to calculate C and A and see how the print() functions create some nice printouts.

In [ ]:
# Import the math package
import math

# Definition of radius
r = 0.43

# Calculate C
C = 2 * math.pi * r

# Calculate A
A = math.pi * r ** 2

# Build printout
print("Circumference: " + str(C))
print("Area: " + str(A))

#### Selective import
General imports, like import math, make all functionality from the math package available to you. However, if you decide to only use a specific part of a package, you can always make your import more selective:

from math import pi

Let's say the Moon's orbit around planet Earth is a perfect circle, with a radius r (in km) that is defined in the script.

In [42]:
# Import radians function of math package
from math import radians

# Definition of radius
r = 192500

# Travel distance of Moon over 12 degrees. Store in dist.
dist = r * radians(12)


# Print out dist

print(dist)


40317.10572106901
